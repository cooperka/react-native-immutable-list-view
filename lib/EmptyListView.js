Object.defineProperty(exports,"__esModule",{value:true});var _extends=Object.assign||function(target){for(var i=1;i<arguments.length;i++){var source=arguments[i];for(var key in source){if(Object.prototype.hasOwnProperty.call(source,key)){target[key]=source[key];}}}return target;};var _jsxFileName='src/EmptyListView.js';var _createClass=function(){function defineProperties(target,props){for(var i=0;i<props.length;i++){var descriptor=props[i];descriptor.enumerable=descriptor.enumerable||false;descriptor.configurable=true;if("value"in descriptor)descriptor.writable=true;Object.defineProperty(target,descriptor.key,descriptor);}}return function(Constructor,protoProps,staticProps){if(protoProps)defineProperties(Constructor.prototype,protoProps);if(staticProps)defineProperties(Constructor,staticProps);return Constructor;};}();var _immutable=require('immutable');var _immutable2=_interopRequireDefault(_immutable);var _propTypes=require('prop-types');var _propTypes2=_interopRequireDefault(_propTypes);var _react=require('react');var _react2=_interopRequireDefault(_react);var _reactNative=require('react-native');var _ImmutableListView=require('./ImmutableListView');var _ImmutableListView2=_interopRequireDefault(_ImmutableListView);function _interopRequireDefault(obj){return obj&&obj.__esModule?obj:{default:obj};}function _classCallCheck(instance,Constructor){if(!(instance instanceof Constructor)){throw new TypeError("Cannot call a class as a function");}}function _possibleConstructorReturn(self,call){if(!self){throw new ReferenceError("this hasn't been initialised - super() hasn't been called");}return call&&(typeof call==="object"||typeof call==="function")?call:self;}function _inherits(subClass,superClass){if(typeof superClass!=="function"&&superClass!==null){throw new TypeError("Super expression must either be null or a function, not "+typeof superClass);}subClass.prototype=Object.create(superClass&&superClass.prototype,{constructor:{value:subClass,enumerable:false,writable:true,configurable:true}});if(superClass)Object.setPrototypeOf?Object.setPrototypeOf(subClass,superClass):subClass.__proto__=superClass;}var styles=_reactNative.StyleSheet.create({centerContent:{alignItems:'center',justifyContent:'center'},emptyText:{margin:12,fontSize:14}});var EmptyListView=function(_PureComponent){_inherits(EmptyListView,_PureComponent);function EmptyListView(){var _ref;var _temp,_this,_ret;_classCallCheck(this,EmptyListView);for(var _len=arguments.length,args=Array(_len),_key=0;_key<_len;_key++){args[_key]=arguments[_key];}return _ret=(_temp=(_this=_possibleConstructorReturn(this,(_ref=EmptyListView.__proto__||Object.getPrototypeOf(EmptyListView)).call.apply(_ref,[this].concat(args))),_this),_this.state={listData:_immutable2.default.List([''])},_temp),_possibleConstructorReturn(_this,_ret);}_createClass(EmptyListView,[{key:'renderRow',value:function renderRow(){var emptyText=this.props.emptyText;return _react2.default.createElement(_reactNative.View,{style:styles.centerContent,__source:{fileName:_jsxFileName,lineNumber:57}},_react2.default.createElement(_reactNative.Text,{style:styles.emptyText,__source:{fileName:_jsxFileName,lineNumber:58}},emptyText));}},{key:'render',value:function render(){var _this2=this;var listData=this.state.listData;return _react2.default.createElement(_ImmutableListView2.default,_extends({immutableData:listData,renderRow:function renderRow(){return _this2.renderRow();}},this.props,{__source:{fileName:_jsxFileName,lineNumber:69}}));}}]);return EmptyListView;}(_react.PureComponent);EmptyListView.propTypes=_extends({},_reactNative.ListView.propTypes,{dataSource:_propTypes2.default.oneOf([undefined]),renderRow:_propTypes2.default.func,emptyText:_propTypes2.default.string});EmptyListView.defaultProps=_extends({},_reactNative.ListView.defaultProps,{emptyText:'No data.'});exports.default=EmptyListView;