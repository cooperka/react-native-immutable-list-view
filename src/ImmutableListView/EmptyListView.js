import Immutable from 'immutable';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import { Text, ListView } from 'react-native';

import ImmutableListView from './ImmutableListView';

import styles from '../styles';
import utils from '../utils';

/**
 * A ListView that displays a single item showing that there is nothing to display.
 * Useful e.g. for preserving the ability to pull-refresh an empty list.
 */
class EmptyListView extends PureComponent {

  static propTypes = {
    // Pass through any props that ListView would normally take.
    ...ListView.propTypes,

    // ImmutableListView handles creating the dataSource, so don't allow it to be passed in.
    dataSource: PropTypes.oneOf([undefined]),

    // Make this prop optional instead of required.
    renderRow: PropTypes.func,

    emptyText: PropTypes.string,
  };

  static defaultProps = {
    ...ListView.defaultProps,

    emptyText: 'No data.',
  };

  state = {
    listData: utils.UNITARY_LIST,
  };

  componentWillMount() {
    this.setListDataFromProps(this.props);
  }

  componentWillReceiveProps(nextProps) {
    this.setListDataFromProps(nextProps);
  }

  setListDataFromProps(props) {
    const { listData } = this.state;
    const { renderEmpty, renderEmptyInList, emptyText } = props;

    // Update the data to make sure the list re-renders if any of the relevant props have changed.
    this.setState({
      listData: listData.set(0, Immutable.fromJS([renderEmpty, renderEmptyInList, emptyText])),
    });
  }

  /**
   * Returns a simple text element showing the `emptyText` string.
   * This method can be overridden by passing in your own `renderRow` prop instead.
   */
  renderRow() {
    const { emptyText } = this.props;

    return (
      <Text style={styles.emptyText}>
        {emptyText}
      </Text>
    );
  }

  render() {
    const { listData } = this.state;
    const {
      renderEmpty, renderEmptyInList, renderSectionHeader, emptyText, ...passThroughProps
    } = this.props;

    return (
      <ImmutableListView
        renderRow={() => this.renderRow()}
        {...passThroughProps}
        immutableData={listData}
      />
    );
  }

}

export { EmptyListView };
