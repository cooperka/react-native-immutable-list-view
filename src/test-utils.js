/* eslint-disable import/no-extraneous-dependencies */
import Immutable from 'immutable';
import React from 'react';
import { Text, ListView } from 'react-native';
import renderer from 'react-test-renderer';
/* eslint-enable */

import ImmutableListView from './ImmutableListView';
import ImmutableVirtualizedList from './ImmutableVirtualizedList';

/**
 * Some common types of data you may want to render with ImmutableListView.
 * @see https://facebook.github.io/react-native/docs/listviewdatasource.html#constructor
 */
const data = {
  EMPTY_DATA: Immutable.List(),

  LIST_DATA: Immutable.List(['lists', 'are', 'great']),

  LIST_DATA_NESTED: Immutable.List([['so', 'are'], ['nested', 'lists']]),

  MAP_DATA_LIST_ROWS: Immutable.fromJS({
    first: ['m', 'a', 'p'],
    second: ['foo'],
    third: [],
    fourth: ['bar'],
  }),

  MAP_DATA_MAP_ROWS: Immutable.fromJS({
    first: {
      row1: 'data 1',
      row2: 'data 2',
    },
    second: {},
  }),

  SET_DATA: Immutable.Set(['one', 'two', 'three']),

  RANGE_DATA: Immutable.Range(3, 10, 3),
};

const renderers = {
  /**
   * @param {*} rowData
   */
  renderRow(rowData) {
    return <Text>{JSON.stringify(rowData)}</Text>;
  },

  // eslint-disable-next-line react/prop-types
  renderItem({ item }) {
    return <Text>{JSON.stringify(item)}</Text>;
  },

  /**
   * @param {Immutable.Iterable} sectionData
   * @param {String} category
   */
  renderSectionHeader(sectionData, category) {
    return <Text header>{`${category} (${sectionData.size} items)`}</Text>;
  },
};

const mocks = {
  /**
   * Mock ScrollView so that it doesn't contain any props when rendered by ListView.
   * This is useful for comparison between ListView and ImmutableListView.
   *
   * @returns {ImmutableListView}
   */
  getImmutableListViewWithoutProps() {
    jest.resetModules();

    // eslint-disable-next-line react/prop-types
    const mockScrollView = props =>
      React.createElement('ScrollView', {}, props.children);
    jest.doMock('ScrollView', () => mockScrollView);

    // eslint-disable-next-line global-require
    return require('./ImmutableListView').default;
  },
};

const expectors = {
  expectToMatchSnapshotWithData(immutableData, shouldRenderSectionHeaders) {
    const renderSectionHeaderProps = shouldRenderSectionHeaders
      ? { renderSectionHeader: renderers.renderSectionHeader }
      : {};

    const tree = renderer
      .create(
        <ImmutableListView
          immutableData={immutableData}
          renderRow={renderers.renderRow}
          {...renderSectionHeaderProps}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  },

  expectVirtualizedToMatchSnapshotWithData(immutableData) {
    const tree = renderer
      .create(
        <ImmutableVirtualizedList
          immutableData={immutableData}
          renderItem={renderers.renderItem}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  },

  expectToMatchListViewWithData(immutableData, shouldRenderSectionHeaders) {
    const MockedImmutableListView = mocks.getImmutableListViewWithoutProps();

    const dataSource = new ListView.DataSource({
      rowHasChanged: (r1, r2) => r1 !== r2,
      sectionHeaderHasChanged: (s1, s2) => s1 !== s2,
    });

    const renderSectionHeaderProps = shouldRenderSectionHeaders
      ? { renderSectionHeader: renderers.renderSectionHeader }
      : {};

    const immutableTree = renderer
      .create(
        <MockedImmutableListView
          immutableData={immutableData}
          renderRow={renderers.renderRow}
          renderEmptyInList={null}
          {...renderSectionHeaderProps}
        />
      )
      .toJSON();

    const updatedDataSource = dataSource.cloneWithRows(immutableData.toJS());
    const regularTree = renderer
      .create(
        <ListView
          dataSource={updatedDataSource}
          renderRow={renderers.renderRow}
          {...renderSectionHeaderProps}
        />
      )
      .toJSON();

    expect(immutableTree).toEqual(regularTree);
  },
};

export { data, renderers, mocks, expectors };
