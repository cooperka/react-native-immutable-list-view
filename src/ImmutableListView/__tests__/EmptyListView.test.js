import React from 'react';
import renderer from 'react-test-renderer';

import { renderers } from '../../test-utils';

import { EmptyListView } from '../EmptyListView';

describe('EmptyListView', () => {
  it('renders with default text', () => {
    const tree = renderer.create(
      <EmptyListView />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders with custom text', () => {
    const tree = renderer.create(
      <EmptyListView
        emptyText="Nothing. Nothing at all."
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders with custom renderRow', () => {
    const tree = renderer.create(
      <EmptyListView
        emptyText="Nothing. Nothing at all."
        renderRow={() => renderers.renderRow('Overridden!')}
      />,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
