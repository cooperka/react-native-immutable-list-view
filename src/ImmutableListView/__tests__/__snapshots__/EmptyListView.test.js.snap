// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyListView renders with custom renderRow 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 1,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      "Overridden!"
    </Text>
  </View>
</RCTScrollView>
`;

exports[`EmptyListView renders with custom text 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 1,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "padding": 8,
          "textAlign": "center",
        }
      }
    >
      Nothing. Nothing at all.
    </Text>
  </View>
</RCTScrollView>
`;

exports[`EmptyListView renders with default text 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={true}
  dataSource={
    ListViewDataSource {
      "items": 1,
    }
  }
  enableEmptySections={true}
  initialListSize={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={1000}
  onKeyboardDidHide={undefined}
  onKeyboardDidShow={undefined}
  onKeyboardWillHide={undefined}
  onKeyboardWillShow={undefined}
  onLayout={[Function]}
  onScroll={[Function]}
  pageSize={1}
  removeClippedSubviews={false}
  renderRow={[Function]}
  scrollEventThrottle={50}
  scrollRenderAheadDistance={1000}
  stickyHeaderIndices={Array []}
  stickySectionHeadersEnabled={true}
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "padding": 8,
          "textAlign": "center",
        }
      }
    >
      No data.
    </Text>
  </View>
</RCTScrollView>
`;
